File: src/http.ts
// ─────────────────────────────────────────────────────────────
//  DSFlix SDK  –  src/http.ts
//  Internal HTTP request wrapper with timeout, auth, and error
// ─────────────────────────────────────────────────────────────

import { DsfError } from './error';
import { DsfClientConfig, DsfErrorResponse } from './types';

export class HttpClient {
  private config: Required<DsfClientConfig>;

  constructor(config: DsfClientConfig) {
    this.config = {
      baseUrl: 'https://api.dawensflix.com',
      timeout: 10000,
      language: 'en-US',
      ...config,
    };
  }

  private buildUrl(path: string, params?: Record<string, string | number | boolean | undefined>): string {
    const url = new URL(`${this.config.baseUrl}${path}`);
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
          url.searchParams.set(key, String(value));
        }
      });
    }
    return url.toString();
  }

  async get<T>(path: string, params?: Record<string, string | number | boolean | undefined>): Promise<T> {
    const url = this.buildUrl(path, params);
    return this.request<T>(url, { method: 'GET' });
  }

  async post<T>(path: string, body?: unknown): Promise<T> {
    const url = this.buildUrl(path);
    return this.request<T>(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: body ? JSON.stringify(body) : undefined,
    });
  }

  async put<T>(path: string, body?: unknown): Promise<T> {
    const url = this.buildUrl(path);
    return this.request<T>(url, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: body ? JSON.stringify(body) : undefined,
    });
  }

  async delete<T>(path: string): Promise<T> {
    const url = this.buildUrl(path);
    return this.request<T>(url, { method: 'DELETE' });
  }

  private async request<T>(url: string, init: RequestInit): Promise<T> {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.config.timeout);

    const headers: Record<string, string> = {
      'Authorization': `Bearer ${this.config.apiKey}`,
      'Accept': 'application/json',
      'X-DSFlix-SDK': 'dsflix-sdk/1.0.0',
      ...(init.headers as Record<string, string> || {}),
    };

    try {
      const response = await fetch(url, {
        ...init,
        headers,
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      const contentType = response.headers.get('content-type') || '';
      const isJson = contentType.includes('application/json');
      const data = isJson ? await response.json() : await response.text();

      if (!response.ok) {
        const errData = data as DsfErrorResponse;
        throw new DsfError(
          errData?.error || `Request failed with status ${response.status}`,
          response.status,
          errData?.detail,
        );
      }

      return data as T;
    } catch (err) {
      clearTimeout(timeoutId);
      if (err instanceof DsfError) throw err;
      if ((err as Error).name === 'AbortError') {
        throw new DsfError('Request timed out', 408);
      }
      throw new DsfError((err as Error).message || 'Network error', 0);
    }
  }
}

File: python/dsflix_sdk/http.py
import requests
from requests import Response

from .exceptions import DsfError


class HttpClient:
    """Internal HTTP wrapper for DSFlix API requests."""

    def __init__(
        self,
        api_key: str,
        base_url: str = "https://api.dawensflix.com",
        timeout: int = 10,
        language: str = "en-US",
    ) -> None:
        self.api_key = api_key
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.language = language
        self.session = requests.Session()
        self.session.headers.update(
            {
                "Authorization": f"Bearer {self.api_key}",
                "Accept": "application/json",
                "X-DSFlix-SDK": "dsflix-sdk-python/0.1.0",
            }
        )

    def get(self, path: str, params: dict | None = None) -> dict:
        return self._request("GET", path, params=params)

    def post(self, path: str, json: dict | None = None, files: dict | None = None) -> dict:
        return self._request("POST", path, json=json, files=files)

    def put(self, path: str, json: dict | None = None) -> dict:
        return self._request("PUT", path, json=json)

    def delete(self, path: str) -> dict:
        return self._request("DELETE", path)

    def _request(
        self,
        method: str,
        path: str,
        params: dict | None = None,
        json: dict | None = None,
        files: dict | None = None,
    ) -> dict:
        url = f"{self.base_url}{path}"

        try:
            response: Response = self.session.request(
                method=method,
                url=url,
                params={k: v for k, v in (params or {}).items() if v is not None},
                json=json,
                files=files,
                timeout=self.timeout,
            )
        except requests.exceptions.Timeout as exc:
            raise DsfError("Request timed out", 408) from exc
        except requests.exceptions.RequestException as exc:
            raise DsfError(str(exc), 0) from exc

        return self._handle_response(response)

    def _handle_response(self, response: Response) -> dict:
        if response.headers.get("Content-Type", "").startswith("application/json"):
            payload = response.json()
        else:
            payload = {"message": response.text}

        if not response.ok:
            error_message = payload.get("error") or payload.get("message") or response.reason
            raise DsfError(error_message, response.status_code, payload.get("detail"))

        return payload
